client/txn.go

```diff
@@ -384,47 +378,40 @@ func (txn *Txn) exec(retryable func(txn *Txn) error) error {
 // always commit or clean-up explicitly even when that may not be
 // required (or even erroneous).
 func (txn *Txn) send(calls ...proto.Call) error {
+	if txn.Proto.Status != proto.PENDING {
+		return util.Errorf("attempting to use %s transaction", txn.Proto.Status)
+	}
+
 	lastIndex := len(calls) - 1
 	if lastIndex < 0 {
 		return nil
 	}
 
+	lastReq := calls[lastIndex].Args
 	// haveTxnWrite tracks intention to write. This is in contrast to
 	// txn.Proto.Writing, which is set by the coordinator when the first
 	// intent has been created, and which lives for the life of the
 	// transaction.
-	var haveTxnWrite bool
-	var haveEndTxn bool
+	haveTxnWrite := proto.IsTransactionWrite(lastReq)
 
-	for i, call := range calls {
+	for _, call := range calls[:lastIndex] {
 		request := call.Args
 
-		if !haveTxnWrite {
-			haveTxnWrite = proto.IsTransactionWrite(request)
+		if req, ok := request.(*proto.EndTransactionRequest); ok {
+			return util.Errorf("%s sent as non-terminal call", req.Method())
 		}
 
-		if req, ok := request.(*proto.EndTransactionRequest); ok {
-			if i != lastIndex {
-				return util.Errorf("%s sent as non-terminal call", req.Method())
-			}
-			haveEndTxn = true
+		if !haveTxnWrite {
+			haveTxnWrite = proto.IsTransactionWrite(request)
 		}
 	}
 
-	// If the transaction record indicates that the coordinator never wrote
-	// an intent (and the client doesn't have one lined up), then there's no
-	// need to send EndTransaction. If there is one anyways, cut it off.
-	if haveEndTxn && (txn.finishedOK || !(txn.Proto.Writing || haveTxnWrite)) {
-		calls = calls[:lastIndex]
-	}
+	_, haveEndTxn := lastReq.(*proto.EndTransactionRequest)
+	needEndTxn := txn.Proto.Writing || haveTxnWrite
 
-	if err := txn.db.send(calls...); err != nil {
-		return err
+	if haveEndTxn && !needEndTxn {
+		calls = calls[:lastIndex]
```

Comment:
`TxnCoordSender` makes no decisions about `EndTransaction`. It's always the client's chore and consequently it must send the call when it's required. On the other hand, the client never gets to change the txn proto except through an update received with a reply or an error. That we have to do it is specifically a hack required because of the way we use `Txn` from `sql`, and so I'm not sure why it should be in the client.

Generally `EndTransaction` handling could be easier if `TxnCoordSender` somehow dropped the call or returned a dummy answer, but I think that's a recipe for more confusion. If we want to distinguish between read-write and read-only transactions (and we should, it saves a lot of work) it also makes sense to be as explicit as possible about the calls, only sending them when needed which gives us the chance to error out on anything that looks wrong.